KeyError during model_rebuild/recursive schemas

### Initial Checks

- [X] I confirm that I'm using Pydantic V2 installed directly from the `main` branch, or equivalent

### Description

Defining a recursive model via model_rebuild _sometimes_ fails with:

```
s = {'schema_ref': 'aiopenapi3.me.L8235d4c12_d58c_4316_829f_4f72825f8720_:94308158795872', 'type': 'definition-ref'}
recurse = <bound method _WalkCoreSchema._walk of <pydantic._internal._core_utils._WalkCoreSchema object at 0x7f8d34350a60>>

    def count_refs(s: core_schema.CoreSchema, recurse: Recurse) -> core_schema.CoreSchema:
        if not is_definition_ref_schema(s):
            return recurse(s, count_refs)
        ref = s['schema_ref']
        ref_counts[ref] += 1
        if current_recursion_ref_count[ref] != 0:
            involved_in_recursion[ref] = True
            return s
    
        current_recursion_ref_count[ref] += 1
>       recurse(all_defs[ref], count_refs)
E       KeyError: 'aiopenapi3.me.L8235d4c12_d58c_4316_829f_4f72825f8720_:94308158795872'
```

So far I can not reproduce easily, but I saw `recent changes` to _simplify_schema_references so … maybe … this is enough already?


### Example Code

_No response_

### Python, Pydantic & OS Version

```Text
pydantic version: 2.0a4
        pydantic-core version: 0.30.0 release build profile
                 install path: ~/venv/openapi3/lib/python3.10/site-packages/pydantic
               python version: 3.10.6 (main, Mar 10 2023, 10:55:28) [GCC 11.3.0]
                     platform: Linux-5.19.0-40-generic-x86_64-with-glibc2.35
     optional deps. installed: ['email-validator', 'typing-extensions']
```

--- a/tests/test_forward_ref.py
+++ b/tests/test_forward_ref.py
@@ -957,3 +957,37 @@ def test_undefined_types_warning_raised_by_usage(create_module):
                 a: UndefinedType
 
             Foobar(a=1)
+
+
+def test_rebuild_recursive_schema():
+    from typing import ForwardRef, List
+
+    class Expressions_(BaseModel):
+        model_config = dict(undefined_types_warning=False)
+        items: List["types['Expression']"]
+
+    class Expression_(BaseModel):
+        model_config = dict(undefined_types_warning=False)
+        Or: ForwardRef("types['allOfExpressions']")
+        Not: ForwardRef("types['allOfExpression']")
+
+    class allOfExpression_(BaseModel):
+        model_config = dict(undefined_types_warning=False)
+        Not: ForwardRef("types['Expression']")
+
+    class allOfExpressions_(BaseModel):
+        model_config = dict(undefined_types_warning=False)
+        items: List["types['Expression']"]
+
+    types_namespace = {
+        'types': {
+            'Expression': Expression_,
+            'Expressions': Expressions_,
+            'allOfExpression': allOfExpression_,
+            'allOfExpressions': allOfExpressions_,
+        }
+    }
+
+    models = [allOfExpressions_, Expressions_]
+    for m in models:
+        m.model_rebuild(_types_namespace=types_namespace)

